### PR TITLE
add HTTP Client tracing

### PR DESCRIPTION
**Pull request by LordSimal**
image

### CODE REVIEW

**Comment by LordSimal** [comment]
I am confused - how does sentry know how long the request took and/or when the span should be started? It looks to me like it starts the span at the end of when the response was already received.

File: src/Http/CakeHttpEventListener.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        $request = $event->getRequest();
+        $response = $event->getResult();
+
+        $parentSpan = SentrySdk::getCurrentHub()->getSpan();
+
+        if ($parentSpan !== null) {
+            $context = SpanContext::make()
+                ->setOp('http.client');
+            $span = $parentSpan->startChild($context);
```
Comment (cleptric) [inline src/Http/CakeHttpEventListener.php:38]: You need to start the span on `HttpClient.beforeSend` and end it on `HttpClient.afterSend`.

File: src/Http/CakeHttpEventListener.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            $uri = $request->getUri();
+            $fullUrl = $uri->getScheme() . '://' . $uri->getHost() . $uri->getPath() . '/';
+            $span
+                ->setDescription(sprintf('%s %s', $request->getMethod(), $fullUrl))
+                ->setStatus(SpanStatus::createFromHttpStatusCode($response?->getStatusCode() ?? 0))
+                ->setData([
+                    'http.request.method' => $request->getMethod(),
+                    'http.response.body.size' => $response?->getBody()?->getSize(),
+                    'http.response.status_code' => $response?->getStatusCode() ?? 0,
+                ])
```
Comment (cleptric) [inline src/Http/CakeHttpEventListener.php:52]: https://develop.sentry.dev/sdk/performance/modules/requests/ for all supported data attributes.

**Comment by cleptric** [comment]
Have a look at https://github.com/getsentry/sentry-laravel/blob/master/src/Sentry/Laravel/Features/HttpClientIntegration.php for some inspiration.

**Comment by LordSimal** [comment]
thanks, that definitely helped.

File: src/Middleware/CakeSentryPerformanceMiddleware.php [hunk 3]
```diff
@@ -76,6 +77,7 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
         $this->addQueryData();
         $listener = new EventListener();
         EventManager::instance()->on($listener);
+        EventManager::instance()->on(new HttpEventListener());
```
Comment (LordSimal) [inline src/Middleware/CakeSentryPerformanceMiddleware.php:80]: In the next major realease of this plugin I would like to unify the listener classes into their own folder/namespace. Didn't expect that I already need multiple implementations of different event listeners already.

**Pull request merged**